libraries/chain/webassembly/eos-vm-oc/LLVMEmitIR.cpp

```diff
@@ -0,0 +1,1303 @@
+			push(irBuilder.CreateLoad(localPointers[imm.variableIndex]));
+		}
+		void set_local(GetOrSetVariableImm<false> imm)
+		{
+			WAVM_ASSERT_THROW(imm.variableIndex < localPointers.size());
+			auto value = irBuilder.CreateBitCast(pop(),localPointers[imm.variableIndex]->getType()->getPointerElementType());
+			irBuilder.CreateStore(value,localPointers[imm.variableIndex]);
+		}
+		void tee_local(GetOrSetVariableImm<false> imm)
+		{
+			WAVM_ASSERT_THROW(imm.variableIndex < localPointers.size());
+			auto value = irBuilder.CreateBitCast(getTopValue(),localPointers[imm.variableIndex]->getType()->getPointerElementType());
+			irBuilder.CreateStore(value,localPointers[imm.variableIndex]);
+		}
+		
+		void get_global(GetOrSetVariableImm<true> imm)
+		{
+			WAVM_ASSERT_THROW(imm.variableIndex < moduleContext.globals.size());
+			if(moduleContext.globals[imm.variableIndex]->getType()->isPointerTy())
+				push(irBuilder.CreateLoad(moduleContext.globals[imm.variableIndex]));
+			else
+				push(moduleContext.globals[imm.variableIndex]);
+		}
+		void set_global(GetOrSetVariableImm<true> imm)
+		{
+			WAVM_ASSERT_THROW(imm.variableIndex < moduleContext.globals.size());
+			auto value = irBuilder.CreateBitCast(pop(),moduleContext.globals[imm.variableIndex]->getType()->getPointerElementType());
+			irBuilder.CreateStore(value,moduleContext.globals[imm.variableIndex]);
+		}
+
+		//
+		// Memory size operators
+		// These just call out to wavmIntrinsics.growMemory/currentMemory, passing a pointer to the default memory for the module.
+		//
+
+		void grow_memory(MemoryImm)
+		{
+			auto deltaNumPages = pop();
+			auto maxMemoryPages = emitLiteral((U32)moduleContext.module.memories.defs[0].type.size.max);
+			auto previousNumPages = emitRuntimeIntrinsic(
+				"eosvmoc_internal.grow_memory",
+				FunctionType::get(ResultType::i32,{ValueType::i32,ValueType::i32}),
+				{deltaNumPages,maxMemoryPages});
+			push(previousNumPages);
+		}
+		void current_memory(MemoryImm)
+		{
+			auto offset = emitLiteral((I32)-18880); ///XXX literal
+			auto bytePointer = CreateInBoundsGEPWAR(irBuilder, moduleContext.defaultMemoryBase, offset);
+			auto ptrTo = irBuilder.CreatePointerCast(bytePointer,llvmI32Type->getPointerTo(256));
+			auto load = irBuilder.CreateLoad(ptrTo);
+			push(load);
+		}
+
+		//
+		// Constant operators
+		//
+
+		#define EMIT_CONST(typeId,nativeType) void typeId##_const(LiteralImm<nativeType> imm) { push(emitLiteral(imm.value)); }
+		EMIT_CONST(i32,I32) EMIT_CONST(i64,I64)
+		EMIT_CONST(f32,F32) EMIT_CONST(f64,F64)
+
+		//
+		// Load/store operators
+		//
+
+		#define EMIT_LOAD_OP(valueTypeId,name,llvmMemoryType,naturalAlignmentLog2,conversionOp) \
+			void valueTypeId##_##name(LoadOrStoreImm<naturalAlignmentLog2> imm) \
+			{ \
+				auto byteIndex = pop(); \
+				auto pointer = coerceByteIndexToPointer(byteIndex,imm.offset,llvmMemoryType); \
+				auto load = irBuilder.CreateLoad(pointer); \
+				load->setAlignment(1<<imm.alignmentLog2); \
```

Comment:
llvm load docs: "Overestimating the alignment results in undefined behavior"

WebAssembly Spec: "The alignment memarg.align in load and store instructions does
not affect the semantics... Unaligned access violating that property is still
allowed and must succeed regardless of the annotation"